lib/chef/provider_resolver.rb

```diff
@@ -68,27 +68,39 @@ def provided_by?(provider_class)
       potential_handlers.include?(provider_class)
     end
 
-    def self.includes_handler?(resource_name, provider_class)
-      priority_map.includes_handler?(resource_name, provider_class)
-    end
-
     def enabled_handlers
-      @enabled_handlers ||=
-        potential_handlers.select { |handler| !overrode_provides?(handler) || handler.provides?(node, resource) }
+      potential_handlers.select { |handler| !overrode_provides?(handler) || handler.provides?(node, resource) }
     end
 
     # TODO deprecate this and allow actions to be passed as a filter to
     # `provides` so we don't have to have two separate things.
     # @api private
     def supported_handlers
-      @supported_handlers ||=
-        enabled_handlers.select { |handler| handler.supports?(resource, action) }
+      enabled_handlers.select { |handler| handler.supports?(resource, action) }
     end
 
     private
 
     def potential_handlers
-      priority_map.list_handlers(node, resource.resource_name)
+      handler_map.list(node, resource.resource_name).uniq
+    end
+
+    # The list of handlers, with any in the priority_map moved to the front
+    def prioritized_handlers
+      @prioritized_handlers ||= begin
+        supported_handlers = self.supported_handlers
+        if supported_handlers.empty?
+          # if none of the providers specifically support the resource, we still need to pick one of the providers that are
+          # enabled on the node to handle the why-run use case. FIXME we should only do this in why-run mode then.
+          Chef::Log.debug "No providers responded true to `supports?` for action #{action} on resource #{resource}, falling back to enabled handlers so we can return something anyway."
+          supported_handlers = enabled_handlers
```

Comment:
This case causes work that's been done before to be done again. Not sure if it will have any measurable performance impact